wurst/file/FileIO.wurst

```diff
@@ -0,0 +1,105 @@
+package FileIO
+import public ChunkedString
+import AbilityIds
+import ErrorHandling
+
+/* 	FileIO for Patches 1.31 and up with chunked string support for big data.
+	Reads and writes using preload files by manipulating default ability tooltips.
+
+	Read a file:
+	>	let fileContent = new File("MyFileName.pld").readAsString()
+
+	Write a file:
+	>	new File("MyFileName.pld")..write("MyContent")..close()
+*/
+
+/** Abilities which tooltips will be used to load data */
+@configurable constant IO_ABILITIES = [AbilityIds.aerialShackles, AbilityIds.rocketAttack, AbilityIds.callToArms,
+						AbilityIds.calltoArms, AbilityIds.cloudofFog, AbilityIds.controlMagic,
+						AbilityIds.defend, AbilityIds.dispelMagic, AbilityIds.feedbackArcaneTower,
+						AbilityIds.feedback]
```

Comment:
Hey. It's very cool that you are doing a new FileIO/etc system with the new natives.
I'm using an upgraded FileIO in my map (still with the old syncing) and I have some suggestions.
There are cases when you may want to insert a thousands of lines in a single file - and the game handles this very well, it's tested! :)
You do not need to use a bunch of abilities, you can use a single one and store data into the tooltips of different levels.
This way you could make the max file size easily configurable.

Like this:
```
@configurable public constant DATA_PACKETS_PER_FILE = 64

constant DATA_DUMMY_ID = 'AXDT'
constant DATA_DUMMY_LEVELS = DATA_PACKETS_PER_FILE

@compiletime function generateDummy()
    let dummy = new AbilityDefinitionIllidanChannel(DATA_DUMMY_ID)
        ..setLevels(DATA_DUMMY_LEVELS)

    for i = 1 to DATA_DUMMY_LEVELS
        dummy.setTooltipNormal(i, "")

constant DATA_DUMMY_RAWCODE = compiletime(DATA_DUMMY_ID.toRawCode())

constant DATA_PER_PRELOAD = 259

constant DATA_PADDING_1 = "\")\r\n\tcall BlzSetAbilityTooltip('" + DATA_DUMMY_RAWCODE + "',\""
constant DATA_PADDING_2 = "\",$"
constant DATA_PADDING_3 = ")\r\n//"
constant DATA_FOOTER = "\" )\r\nendfunction\r\nfunction AAA takes nothing returns nothing \r\n//"

public constant MAX_PACKET_LENGTH = compiletime(DATA_PER_PRELOAD - 
                                                DATA_PADDING_1.length() - 
                                                DATA_PADDING_2.length() - 
                                                DATA_PADDING_3.length() -
                                                DATA_DUMMY_RAWCODE.length() -
												5)
```

And then write it like this:
```
Preload(DATA_PADDING_1 + packet + DATA_PADDING_2 + i.toHex() + DATA_PADDING_3)        
```

Where toHex works like this:
```
function int.toHex(int length) returns string
    var temp = this
    var result = ""
    for i = 1 to length
        result = hexs[temp mod 16] + result
        temp = temp div 16
    return result
```

Easily configurable, no limits, efficient.

Reply:
I mostly copied trigger happy's impl, because that works with 1.29 and 1.30 also, not only 1.31 or so. The multi level thing can be done, but it is still limited by the max string limit, so no, not "no limits" and it makes sense to keep it chunked.

Reply:
Yea. I mean no limits on the total size :)
`i` in the example above was the chunk number.

Reply:
> Yea. I mean no limits on the total size :)

Yes, I ment the total size, which is still limited by the maximum string length.

> A special multilevel ability is also better because you do not need to restore tooltips after reading.

> The suggested method should not also break anything on the previous patches - actually, nothing changes except the ability id and levels.
I can't recall the patch number when I started to use this, but it should be less than 1.31

Before 1.31 you can only read the first ability level, this is why TH did it this way. In 1.29 custom objects can't be read iirc. But I am fine with only supporting 1.31 at this point.

Reply:
> Before 1.31 you can only read the first ability level, this is why TH did it this way. In 1.29 custom objects can't be read iirc.
Ah, okay, didn't know this, thanks for clarifying.

> Yes, I ment the total size, which is still limited by the maximum string length.
Just wondering - do you mean the total file length? As far as I understand, by keeping the strings chunked the total size of the file should not be limited.

Reply:
No, I mean the maximum length of  a Jass `string`. With concatenation it's 4k, which usually should be sufficient, but I wanna keep it flexible enough from the start. Also the 4k string can be annoying to handle, because you can't take big substrings and it's generally a hassle, at least that's what I found from my last tests.

Reply:
Lol, fixed it 😅
As for the length of the strings, I knew this. I was just talking about limits on the file size, which was limited by abilities number.
In my use case I’m writing files with thousands of lines, and the dummy ability has 8k levels 😅

Reply:
Alright, so does anything speak against merging this?

Reply:
This version looks good to me.
Yes, I'm still using the syncing based on gamecache and units selection, playing on 1.31 - it works.

Reply:
hm okay, I just know that it didn't work anymore in EBR, but this sounds like it's better to deprecate instead of delete after all.